doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics, and allow components
+  of a system to each work with a different schema version and merge
+  their results.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+- Managers not required to consider storage/communication of additional
+  version data to enable interpretation by future downstream systems.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Introduces "Try and see" negotiation on the host side, although
+  introspection behaviour to negate this could be introduced at a later
+  date, similar to other introspection mechanisms. (See [In data -
+  Extended](#in-data---extended)).
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would be become a tedious dance.
+- `entityTraits` becomes more complex, as we either have to declare it
+  version agnostic somehow, or managers would need to return the correct
+  versions/entire swathes of versions. (See [What's
+  versioned](#whats-versioned)).
+- Unclear how specification versioning is to work without an out-of-data
+  version.
+- The most obvious implementation, (using `traitID` to store the
+  version), would effectively make any change to a trait version a
+  breaking change. Allowing non breaking changes (such as adding a new
+  property) would require significant rework to the `traitgen` output.
+
+### In data - Extended
+The implementation of in-data versioning is very basic, which is part of
+its strength, but it necessitates a "try and see" approach on the part
+of the host. If the host wants to be adaptive to the manager, it may
+benefit from a more up front introspection of the supported manager
+versions.
+
+To this end, the `supportedSchemaVersions` method (or similar concept,)
+from Option 2 could be orthogonally added without any conceptual
+conflict. This would have the added benefit of allowing hosts to also
+support a range of version/have backwards compatibility, which is a huge
+benefit.
```

Comment:
[question] Any reason that `managementPolicy` can't be used as-is for this?

Reply:
Probably not,  see https://github.com/OpenAssetIO/OpenAssetIO/pull/1267#discussion_r1506772966